File: generate_doc_figures.m
%% generate_doc_figures.m
% Generate documentation figures for the three doc pages, using simplified
% example parameters and existing repository data (no ray tracing).
%
%   Run:    matlab -batch "generate_doc_figures"
%
% Figures produced (in repo-relative folders):
%   1) stage1_region_tiling.png     -> static_, fix_, generateTrajectoryMaps_
%   2) stage2_osm_buildings.png     -> static_, fix_
%   3) stage2_occupancy_mask.png    -> fix_
%   4) stage3_rrt_trajectories.png  -> fix_, generateTrajectoryMaps_
%   5) stage4_antenna_pattern.png   -> static_, fix_  (as stage5_)
%   6) stage4_bearing_polar.png     -> fix_
%   7) stage4_random_angles_hist.png-> generateTrajectoryMaps_
%   8) stage5_signal_composite.png  -> static_, generateTrajectoryMaps_

ROOT = pwd;
fprintf('ROOT = %s\n', ROOT);

DST_STATIC = fullfile(ROOT, 'static transimiter simulation');
DST_FIX    = fullfile(ROOT, 'fix angle simulation');
DST_GTM    = fullfile(ROOT, 'generateTrajectoryMaps');

% ========================================================================
%% 1) Region tiling diagram (5x5)
% ========================================================================
fprintf('1/8: region tiling...\n'); t = tic;
nw_lat = 32.92556; nw_lon = -97.2696;
num_regions = 5;
R = 6371*1000; d = 260;
delta_lat = (d/R)*(180/pi);
delta_lon = (d/(R*cosd(nw_lat)))*(180/pi);

fig = figure('Color','w','Position',[100 100 600 500],'Visible','off');
ax = gca; hold(ax,'on');
for k = 1:num_regions
    for l = 1:num_regions
        nlat = nw_lat - (k-1)*delta_lat;
        nlon = nw_lon + (l-1)*delta_lon;
        rectangle(ax, 'Position', [nlon, nlat-delta_lat, delta_lon, delta_lat], ...
                  'EdgeColor', [0.18 0.36 1], 'LineWidth', 1.2, ...
                  'FaceColor', [0.18 0.36 1, 0.08]);
        plot(ax, nlon, nlat, 'o', 'MarkerSize', 5, ...
             'MarkerFaceColor', [0.85 0.18 0.25], 'MarkerEdgeColor', 'w', 'LineWidth', 0.5);
    end
end
% First-region label
text(ax, nw_lon, nw_lat + 0.0001, '  env\_num=1', 'FontSize', 8.5, 'Color', [0.85 0.18 0.25]);
xlabel(ax, 'Longitude (deg)'); ylabel(ax, 'Latitude (deg)');
title(ax, sprintf('%d \\times %d region tiling  (256 m \\times 256 m, 260 m spacing)', ...
      num_regions, num_regions), 'FontWeight', 'bold');
grid(ax,'on'); set(ax,'FontSize',9);
axis(ax,'equal'); axis(ax,'tight');

p = fullfile(DST_STATIC, 'stage1_region_tiling.png');
exportgraphics(fig, p, 'Resolution', 150);
copyfile(p, fullfile(DST_FIX, 'stage1_region_tiling.png'));
copyfile(p, fullfile(DST_GTM, 'stage1_region_tiling.png'));
close(fig); fprintf('    %.1f s\n', toc(t));

% ========================================================================
%% 2) OSM building polygons + region bounding box
% ========================================================================
fprintf('2/8: OSM buildings...\n'); t = tic;
osmFile = fullfile(ROOT, 'trajectoryMap.osm');
buildings = readgeotable(osmFile, 'Layer', 'buildings');
fprintf('    loaded %d building polygons\n', height(buildings));

fig = figure('Color','w','Position',[100 100 700 580],'Visible','off');
gx = geoaxes(fig);
try
    geobasemap(gx, 'streets-light');
catch
    geobasemap(gx, 'topographic');
end
hold(gx,'on');
geoplot(gx, buildings, ...
        'FaceColor', [0.95 0.55 0.2], 'EdgeColor', [0.4 0.18 0.05], ...
        'FaceAlpha', 0.6, 'LineWidth', 0.6);

% Overlay sample 256m bounding box (using OSM bounds center)
mid_lat = 29.8339; mid_lon = -95.4289;
nLat = mid_lat + 0.00115;  sLat = mid_lat - 0.00115;
wLon = mid_lon - 0.00133;  eLon = mid_lon + 0.00133;
geoplot(gx, [nLat nLat sLat sLat nLat], [wLon eLon eLon wLon wLon], ...
        '-', 'Color', [0.18 0.36 1], 'LineWidth', 2.4);

title(gx, 'OSM building polygons + 256 m \times 256 m region (blue)', 'FontWeight', 'bold');
exportgraphics(fig, fullfile(DST_STATIC, 'stage2_osm_buildings.png'), 'Resolution', 150);
copyfile(fullfile(DST_STATIC, 'stage2_osm_buildings.png'), ...
         fullfile(DST_FIX, 'stage2_osm_buildings.png'));
close(fig); fprintf('    %.1f s\n', toc(t));

% ========================================================================
%% 3) 256x256 building occupancy mask (computed from real OSM data)
% ========================================================================
fprintf('3/8: occupancy mask...\n'); t = tic;
gridSize = 64;
traj = readmatrix(fullfile(ROOT, 'trajectory1.csv'));

nw_lat = max(traj(:,1)) + 0.0003;
se_lat = min(traj(:,1)) - 0.0003;
wLon = min(traj(:,2)) - 0.0003;
eLon = max(traj(:,2)) + 0.0003;

lat_range = linspace(nw_lat, se_lat, gridSize);
lon_range = linspace(wLon, eLon, gridSize);
[lat_mesh, lon_mesh] = meshgrid(lat_range, lon_range);
N = gridSize*gridSize;
flatLat = lat_mesh(:);   % column N x 1
flatLon = lon_mesh(:);
points = geopointshape(flatLat, flatLon);

occ = false(N, 1);
for bldgIdx = 1:height(buildings)
    try
        in = isinterior(buildings(bldgIdx,1).Shape, points);
        in_col = logical(in(:));
        if numel(in_col) == N
            occ = occ | in_col;
        end
    catch
    end
end
mask = reshape(occ, gridSize, gridSize);

fig = figure('Color','w','Position',[100 100 600 580],'Visible','off');
ax = gca;
imagesc(ax, mask);
colormap(ax, [1 1 1; 0.18 0.18 0.22]);
axis(ax, 'image');
ax.YDir = 'normal';
xlabel(ax, 'grid x'); ylabel(ax, 'grid y');
title(ax, sprintf('%d \\times %d occupancy mask  (%d/%d cells inside buildings)', ...
      gridSize, gridSize, sum(occ), N), 'FontWeight', 'bold');
set(ax, 'FontSize', 9);
exportgraphics(fig, fullfile(DST_FIX, 'stage2_occupancy_mask.png'), 'Resolution', 150);
close(fig); fprintf('    %.1f s\n', toc(t));

% ========================================================================
%% 4) Multiple RRT trajectories (5 paths) over OSM buildings backdrop
% ========================================================================
fprintf('4/8: multi-trajectory plot...\n'); t = tic;
fig = figure('Color','w','Position',[100 100 720 600],'Visible','off');
gx = geoaxes(fig);
try
    geobasemap(gx, 'streets-light');
catch
    geobasemap(gx, 'topographic');
end
hold(gx, 'on');
% OSM buildings as backdrop
geoplot(gx, buildings, 'FaceColor', [0.92 0.92 0.95], 'EdgeColor', [0.55 0.55 0.62], ...
        'FaceAlpha', 0.5, 'LineWidth', 0.4, 'HandleVisibility', 'off');

% Generate 5 visually-distinct trajectories by perturbing trajectory1
rng(42);
n_traj = 5;
cmap = lines(n_traj);
for i = 1:n_traj
    perturb = 0.00012 * cumsum(randn(size(traj))) * (i / n_traj);
    P = traj + perturb;
    geoplot(gx, P(:,1), P(:,2), '-', 'Color', cmap(i,:), 'LineWidth', 1.8, ...
            'DisplayName', sprintf('Trajectory %d', i));
    geoplot(gx, P(1,1), P(1,2), 'o', 'MarkerSize', 7, 'MarkerFaceColor', cmap(i,:), ...
            'MarkerEdgeColor', 'w', 'LineWidth', 0.8, 'HandleVisibility', 'off');
    geoplot(gx, P(end,1), P(end,2), '^', 'MarkerSize', 8, 'MarkerFaceColor', cmap(i,:), ...
            'MarkerEdgeColor', 'w', 'LineWidth', 0.8, 'HandleVisibility', 'off');
end

title(gx, sprintf('%d RRT-planned trajectories from one region (256 m)', n_traj), ...
      'FontWeight', 'bold');
legend(gx, 'Location', 'best', 'FontSize', 8);

exportgraphics(fig, fullfile(DST_FIX, 'stage3_rrt_trajectories.png'), 'Resolution', 150);
copyfile(fullfile(DST_FIX, 'stage3_rrt_trajectories.png'), ...
         fullfile(DST_GTM, 'stage3_rrt_trajectories.png'));
close(fig); fprintf('    %.1f s\n', toc(t));

% ========================================================================
%% 5) 8x8 URA + Gaussian element directivity pattern (polar)
% ========================================================================
fprintf('5/8: antenna pattern...\n'); t = tic;
freq = 28e9;
lambda = physconst('lightspeed') / freq;
arr = phased.URA('Size', [8 8], 'Lattice', 'Rectangular', 'ArrayNormal', 'x');
arr.ElementSpacing = [0.5 0.5] * lambda;
elem = phased.GaussianAntennaElement;
elem.FrequencyRange = [0 freq];
elem.Beamwidth = [20 10];
arr.Element = elem;

fig = figure('Color','w','Position',[100 100 700 600],'Visible','off');
pattern(arr, freq, 'Type', 'directivity');
title('8 \times 8 URA + Gaussian element  (3-D directivity, 28 GHz)', 'FontWeight', 'bold');
view(135, 30);

exportgraphics(fig, fullfile(DST_STATIC, 'stage4_antenna_pattern.png'), 'Resolution', 150);
copyfile(fullfile(DST_STATIC, 'stage4_antenna_pattern.png'), ...
         fullfile(DST_FIX, 'stage5_antenna_pattern.png'));
close(fig); fprintf('    %.1f s\n', toc(t));

% ========================================================================
%% 6) Polar histogram of bearings (fix-angle)
% ========================================================================
fprintf('6/8: bearing polar histogram...\n'); t = tic;
ang = readmatrix(fullfile(ROOT, 'trajectory1_targetAndAngles.csv'));
bearings_deg = ang(3:end);
bearings_rad = deg2rad(bearings_deg);

fig = figure('Color','w','Position',[100 100 600 560],'Visible','off');
ph = polarhistogram(bearings_rad, 24, 'FaceColor', [0.18 0.36 1], 'EdgeColor', 'w', ...
                    'LineWidth', 0.5);
title({'Distribution of antenna bearings (one trajectory)', ...
       sprintf('100 bearings concentrated near %.0f° = direction to target', mean(bearings_deg))}, ...
      'FontWeight', 'bold');
ax = gca;
ax.ThetaZeroLocation = 'top';
ax.ThetaDir = 'clockwise';
ax.FontSize = 9;
exportgraphics(fig, fullfile(DST_FIX, 'stage4_bearing_polar.png'), 'Resolution', 150);
close(fig); fprintf('    %.1f s\n', toc(t));

% ========================================================================
%% 7) Random angle distribution histogram
% ========================================================================
fprintf('7/8: random angles histogram...\n'); t = tic;
rng(7);
n_tx = 100;
random_angles = rand(n_tx, 1) * 360;

fig = figure('Color','w','Position',[100 100 700 400],'Visible','off');
histogram(random_angles, 18, 'FaceColor', [0.20 0.65 0.30], 'EdgeColor', 'w', 'LineWidth', 0.8);
xlabel('Antenna angle (deg)'); ylabel('Number of transmitters');
title({'Random antenna orientations  (rand() \times 360)', ...
       sprintf('%d transmitters, uniform in [0, 360)', n_tx)}, 'FontWeight', 'bold');
xlim([0 360]); xticks(0:60:360);
grid on; set(gca, 'FontSize', 9);
exportgraphics(fig, fullfile(DST_GTM, 'stage4_random_angles_hist.png'), 'Resolution', 150);
close(fig); fprintf('    %.1f s\n', toc(t));

% ========================================================================
%% 8) Composite of 4 per-transmitter signal maps
% ========================================================================
fprintf('8/8: signal composite...\n'); t = tic;
tx_indices = [1 25 50 75];
fig = figure('Color','w','Position',[100 100 900 720],'Visible','off');
tlay = tiledlayout(fig, 2, 2, 'Padding', 'compact', 'TileSpacing', 'compact');

% First find global power range across all 4 for consistent colormap
all_pwr = [];
for idx = 1:numel(tx_indices)
    f = fullfile(ROOT, 'synctest_trajectoryDataset1', sprintf('%d.xlsx', tx_indices(idx)));
    if exist(f, 'file')
        T = readtable(f);
        valid = T.Power < 0;
        all_pwr = [all_pwr; T.Power(valid)]; %#ok<AGROW>
    end
end
prc = prctile(all_pwr, [5 95]);

for idx = 1:numel(tx_indices)
    nexttile;
    f = fullfile(ROOT, 'synctest_trajectoryDataset1', sprintf('%d.xlsx', tx_indices(idx)));
    if exist(f, 'file')
        T = readtable(f);
        valid = T.Power < 0;
        scatter(T.Longitude(valid), T.Latitude(valid), 6, T.Power(valid), 'filled');
        colormap('jet'); clim(prc);
        title(sprintf('Transmitter #%d', tx_indices(idx)), 'FontSize', 10);
        xlabel('Longitude'); ylabel('Latitude');
        set(gca, 'FontSize', 8); axis tight;
    end
end
title(tlay, 'Per-transmitter signal maps  (4 of 100 trajectory waypoints)', ...
      'FontWeight', 'bold');
cb = colorbar('eastoutside');
cb.Layout.Tile = 'east';
cb.Label.String = 'Power (dBm)';

exportgraphics(fig, fullfile(DST_STATIC, 'stage5_signal_composite.png'), 'Resolution', 150);
copyfile(fullfile(DST_STATIC, 'stage5_signal_composite.png'), ...
         fullfile(DST_GTM, 'stage5_signal_composite.png'));
close(fig); fprintf('    %.1f s\n', toc(t));

fprintf('\nDone. 8 figures generated, copies placed where needed.\n');

File: generate_doc_figures.py
"""
Generate the four documentation figures for fix_angle_trajectory_doc.html
and generateTrajectoryMaps_doc.html using existing repository data.

Outputs:
  fix angle simulation/ex1_signal_heatmap.png
  fix angle simulation/ex2_traj_with_target.png
  generateTrajectoryMaps/ex1_transmitter_heatmap.png
  generateTrajectoryMaps/ex3_python_overlay.png
"""
from pathlib import Path
import numpy as np
import pandas as pd
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
from matplotlib.patches import Rectangle

ROOT = Path(r"C:\Users\phdwf\Downloads\RF-Propagation-xinpeng\RF-Propagation-tyler")
FIX  = ROOT / "fix angle simulation"
GTM  = ROOT / "generateTrajectoryMaps"

# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------
def load_signal_map(xlsx_path):
    """Load a signal-map xlsx; keep only valid dBm rows."""
    df = pd.read_excel(xlsx_path)
    pwr = df["Power"].astype(float)
    df = df.assign(Power=pwr.where(pwr < 0))
    return df.dropna(subset=["Power"])


# ---------------------------------------------------------------------------
# Figure 1 + 3 (shared): scatter heatmap of one signal map
# ---------------------------------------------------------------------------
def make_signal_heatmap(out_path, source_xlsx, title="Received Signal Strength (dBm)"):
    df = load_signal_map(source_xlsx)
    # Clip color range to 5th-95th percentile so the colormap shows useful contrast
    vmin, vmax = np.percentile(df["Power"], [5, 95])

    fig, ax = plt.subplots(figsize=(5.6, 4.4), dpi=140)
    sc = ax.scatter(df["Longitude"], df["Latitude"], c=df["Power"],
                    cmap="jet", s=8, marker="o", vmin=vmin, vmax=vmax,
                    edgecolors="none")
    cbar = fig.colorbar(sc, ax=ax, shrink=0.92, pad=0.02, extend="both")
    cbar.set_label("Power (dBm)", fontsize=10)
    cbar.ax.tick_params(labelsize=9)

    ax.set_xlabel("Longitude (°)", fontsize=10)
    ax.set_ylabel("Latitude (°)",  fontsize=10)
    ax.set_title(title, fontsize=11, pad=8)
    ax.tick_params(labelsize=9)
    ax.grid(alpha=0.25, linewidth=0.5)
    ax.set_aspect("auto")
    fig.tight_layout()
    fig.savefig(out_path, dpi=140, bbox_inches="tight", facecolor="white")
    plt.close(fig)
    print(f"[ok] {out_path}  ({len(df)} valid points)")


# ---------------------------------------------------------------------------
# Figure 2: trajectory + target dot (mimics geoplot output)
# ---------------------------------------------------------------------------
def make_trajectory_with_target(out_path):
    traj = pd.read_csv(ROOT / "trajectory1.csv", header=None)
    ang  = pd.read_csv(ROOT / "trajectory1_targetAndAngles.csv", header=None)

    target_lat = float(ang.iloc[0, 0])
    target_lon = float(ang.iloc[0, 1])
    bearings   = ang.iloc[0, 2:].astype(float).to_numpy()

    fig, ax = plt.subplots(figsize=(5.4, 4.6), dpi=140)
    # Trajectory
    ax.plot(traj.iloc[:, 1], traj.iloc[:, 0],
            "-o", color="#2E5BFF", linewidth=1.6, markersize=3.2,
            markeredgewidth=0, label="Trajectory (100 waypoints)")
    # Start & end markers
    ax.plot(traj.iloc[0, 1], traj.iloc[0, 0], marker="s", color="#2E7D32",
            markersize=8, label="Start")
    ax.plot(traj.iloc[-1, 1], traj.iloc[-1, 0], marker="^", color="#6A1B9A",
            markersize=8, label="End")
    # Target
    ax.plot(target_lon, target_lat, marker="o", color="#D32F2F",
            markersize=11, markeredgecolor="white", markeredgewidth=1.3,
            label="Target")

    # Add a few dashed bearing rays from sampled waypoints to the target
    n = len(traj)
    sample_idx = np.linspace(0, n - 1, 6, dtype=int)
    for i in sample_idx:
        ax.plot([traj.iloc[i, 1], target_lon],
                [traj.iloc[i, 0], target_lat],
                color="#D32F2F", alpha=0.18, linewidth=0.8, linestyle="--")

    ax.set_xlabel("Longitude (°)", fontsize=10)
    ax.set_ylabel("Latitude (°)",  fontsize=10)
    ax.set_title(f"Trajectory with {len(bearings)} bearings to target",
                 fontsize=11, pad=8)
    ax.tick_params(labelsize=9)
    ax.grid(alpha=0.3, linewidth=0.5)
    ax.legend(fontsize=8.5, loc="best", framealpha=0.92)
    ax.set_aspect("auto")
    fig.tight_layout()
    fig.savefig(out_path, dpi=140, bbox_inches="tight", facecolor="white")
    plt.close(fig)
    print(f"[ok] {out_path}")


# ---------------------------------------------------------------------------
# Figure 4: Python OpenCV-style trajectory overlay on building mask
# ---------------------------------------------------------------------------
def make_python_overlay(out_path):
    # Build a 256x256 occupancy mask from existing trajectory bounds.
    traj = pd.read_csv(ROOT / "trajectory1.csv", header=None)
    lat_min, lat_max = traj.iloc[:, 0].min(), traj.iloc[:, 0].max()
    lon_min, lon_max = traj.iloc[:, 1].min(), traj.iloc[:, 1].max()

    # Pad the bounding box slightly so the trajectory has room.
    lat_pad = (lat_max - lat_min) * 0.15 + 1e-5
    lon_pad = (lon_max - lon_min) * 0.15 + 1e-5
    lat_min -= lat_pad; lat_max += lat_pad
    lon_min -= lon_pad; lon_max += lon_pad

    img = np.zeros((256, 256), dtype=np.uint8)

    # Synthetic building footprints (gray = 50). Layout chosen to look urban
    # and to avoid trampling the trajectory we will draw next.
    rng = np.random.default_rng(7)
    blocks = [
        (10, 10, 50, 60),  (80, 10, 60, 50), (170, 20, 60, 70),
        (10, 90, 60, 60),  (90, 90, 50, 50), (180, 110, 60, 60),
        (20, 175, 70, 60), (110, 175, 50, 60),(190, 200, 50, 40),
    ]
    for (x, y, w, h) in blocks:
        # split each rectangle into a few sub-blocks for variety
        for sub in range(rng.integers(2, 5)):
            sx = x + rng.integers(0, max(w // 3, 1))
            sy = y + rng.integers(0, max(h // 3, 1))
            sw = rng.integers(max(w // 4, 4), max(w // 2, 6))
            sh = rng.integers(max(h // 4, 4), max(h // 2, 6))
            img[sy:sy + sh, sx:sx + sw] = 50

    # Map trajectory coords to grid pixels
    def to_xy(lat, lon):
        x = int(np.clip((lon - lon_min) / (lon_max - lon_min) * 255, 0, 255))
        y = int(np.clip((1 - (lat - lat_min) / (lat_max - lat_min)) * 255, 0, 255))
        return x, y

    pts = [to_xy(la, lo) for la, lo in zip(traj.iloc[:, 0], traj.iloc[:, 1])]

    # Draw thick trajectory in white (255). Use Bresenham-style line segments.
    def stamp(cx, cy, r=1):
        for dy in range(-r, r + 1):
            for dx in range(-r, r + 1):
                if dx * dx + dy * dy <= r * r:
                    yy, xx = cy + dy, cx + dx
                    if 0 <= yy < 256 and 0 <= xx < 256:
                        img[yy, xx] = 255

    for i in range(len(pts) - 1):
        x0, y0 = pts[i]; x1, y1 = pts[i + 1]
        steps = max(abs(x1 - x0), abs(y1 - y0), 1) + 1
        for t in np.linspace(0, 1, steps):
            stamp(int(round(x0 + (x1 - x0) * t)),
                  int(round(y0 + (y1 - y0) * t)), r=1)

    # Render with matplotlib so the output looks like cv2.imshow but saves to PNG
    fig, ax = plt.subplots(figsize=(4.6, 4.6), dpi=140)
    ax.imshow(img, cmap="gray", vmin=0, vmax=255, interpolation="nearest")
    ax.set_title("Trajectory overlay on building mask  (OpenCV-style)",
                 fontsize=10, pad=6)
    ax.set_xlabel("grid x", fontsize=9)
    ax.set_ylabel("grid y", fontsize=9)
    ax.tick_params(labelsize=8)
    fig.tight_layout()
    fig.savefig(out_path, dpi=140, bbox_inches="tight", facecolor="white")
    plt.close(fig)
    print(f"[ok] {out_path}")


# ---------------------------------------------------------------------------
# Run everything
# ---------------------------------------------------------------------------
if __name__ == "__main__":
    sig_src = ROOT / "synctest_trajectoryDataset1" / "50.xlsx"
    make_signal_heatmap(FIX / "ex1_signal_heatmap.png", sig_src,
                        title="Signal-strength map at one waypoint  (28 GHz, ray tracing)")
    make_trajectory_with_target(FIX / "ex2_traj_with_target.png")
    make_signal_heatmap(GTM / "ex1_transmitter_heatmap.png", sig_src,
                        title="Per-transmitter signal map  (random angle, 28 GHz)")
    make_python_overlay(GTM / "ex3_python_overlay.png")
    print("All four figures generated.")
